source/hic_hal/nxp/lpc4322/swdp_sgpio.c

```diff
@@ -0,0 +1,1103 @@
+#define SLICE_SHD(s)                             LPC_SGPIO->REG_SS[s]
+
+#define REQUEST_READ(request)                    (request & DAP_TRANSFER_RnW)
+#define REQUEST_TIMESTAMP(request)               (request & DAP_TRANSFER_TIMESTAMP)
+#define REQUEST_WRITE(request)                   ((request & DAP_TRANSFER_RnW) == 0U)
+
+
+
+__STATIC_FORCEINLINE uint8_t PARITY_U32(uint32_t v)
+{
+    v ^= v >> 16;
+    v ^= v >> 8;
+    v ^= v >> 4;
+    v &= 0xF;
+    return (0x6996 >> v) & 1;
+}
+
+
+
+__STATIC_FORCEINLINE void putU32(uint32_t src_sz, const uint8_t *src, volatile uint32_t *dst)
+{
+    uint32_t rv = 0;
+
+    for (unsigned i = 0; i < src_sz; ++i)
+    {
+        rv |= *src++<<(i*8);
+    }
+    *dst = rv;
+}
+
+
+
+__STATIC_FORCEINLINE void getU32(uint32_t src_sz, uint32_t src, uint8_t *dst)
+{
+    for (unsigned i=0; i< src_sz; ++i)
+    {
+        *dst++ = src & 0xFF;
+    }
+}
+
+
+
+// Enable the Output SGPIO pin.
+// 19.6.15 UM10503.
+__STATIC_FORCEINLINE void sgpio_pin_oenable(uint32_t pin)
+{
+    uint32_t cur_oenreg;
+    cur_oenreg = LPC_SGPIO->GPIO_OENREG;
+    LPC_SGPIO->GPIO_OENREG = cur_oenreg | (1 << pin);
+}
+
+
+
+// Disable the Output SGPIO pin.
+// 19.6.15 UM10503.
+__STATIC_FORCEINLINE void sgpio_pin_odisable(uint32_t pin)
+{
+    uint32_t cur_oenreg;
+    cur_oenreg = LPC_SGPIO->GPIO_OENREG;
+    LPC_SGPIO->GPIO_OENREG = cur_oenreg & ~(1 << pin);
+}
+
+
+
+// Exchange clock interrupt clear mask register.
+// 19.6.24 UM10503.
+// Set the CLR_EN_1 register bit to clear
+// the corresponding bit in the ENABLE_1 register.
+__STATIC_FORCEINLINE void exchg_clk_int_off(uint32_t mask)
+{
+    LPC_SGPIO->CLR_EN_1 = mask;
+}
+
+
+
+// Exchange clock interrupt clear status register.
+// 19.6.28 UM10503.
+__STATIC_FORCEINLINE void exchg_clk_cleanup(uint32_t slice_mask)
+{
+    LPC_SGPIO->CLR_STATUS_1 = slice_mask;
+}
+
+
+
+// Exchange clock interrupt status register.
+// 19.6.27 UM10503.
+// The STATUS_1 register bits are set when
+// the shadow and data registers are exchanged.
+__STATIC_FORCEINLINE void exchg_clk_wait_on(uint32_t slice_mask)
+{
+    while ((LPC_SGPIO->STATUS_1 & slice_mask) != slice_mask) { }
+}
+
+
+
+// Enable the slices' COUNT counter.
+// 1 = Starts COUNTn or external shift clock.
+// 0 = Disables slice shift clock.
+__STATIC_FORCEINLINE void shift_clk_kickoff(uint32_t slice_mask)
+{
+    LPC_SGPIO->CTRL_ENABLE |= slice_mask;
+}
+
+
+
+__STATIC_FORCEINLINE void shift_clk_int_off(uint32_t mask)
+{
+    LPC_SGPIO->CLR_EN_0 = mask;
+}
+
+
+
+__STATIC_FORCEINLINE void shift_clk_disable(uint32_t slice_mask)
+{
+    LPC_SGPIO->CTRL_ENABLE &= ~slice_mask;
+}
+
+
+
+// frequency(SHIFT_CLOCK) = frequency(SGPIO_CLK) / (PRESET+1).
+// PRESET = frequency(SGPIO_CLK)/frequency(SHIFT_CLK) - 1.
+// SGPIO_CLK = SystemCoreClock
+// SHIFT_CLOCK = DAP_Data.nominal_clock
+static uint32_t shift_clk_preset()
+{
+    return SystemCoreClock/DAP_Data.nominal_clock - 1;
```

Comment:
This should be using PLL1. This is how I've seen it used in the sgpio_uart NXP's Application Notes.
At the moment I'm trying to make CGU aware of CGU_PERIPHERAL_SGPIO, which is mentioned in the User Manual but
nowhere to be found in lpc43xx_cgu.c.

Reply:
done. using PLL0AUDIO